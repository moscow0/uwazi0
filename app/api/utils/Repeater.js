const timeout = async interval =>
  new Promise(resolve => {
    setTimeout(resolve, interval);
  });

export class Repeater {
  constructor(cb, interval) {
    this.cb = cb;
    this.interval = interval;
    this.stopped = null;
  }

  async start() {
    while (!this.stopped) {
      // eslint-disable-next-line no-await-in-loop
      await this.cb();
      // eslint-disable-next-line no-await-in-loop
      await timeout(this.interval);
    }

    this.stopped();
  }

  async stop() {
    return new Promise(resolve => {
      this.stopped = resolve;
    });
  }
}
