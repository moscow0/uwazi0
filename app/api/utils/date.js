import moment from 'moment';
// @ts-ignore
import parser from 'any-date-parser';

export default {
  currentUTC() {
    return moment.utc().toDate().getTime();
  },

  descriptionToTimestamp(date) {
    if (date === 'last-day-last-month') {
      return moment.utc().subtract(1, 'months').endOf('month').unix();
    }

    if (date === 'first-day-last-month') {
      return moment.utc().subtract(1, 'months').startOf('month').unix();
    }

    return date;
  },

  addYearsToCurrentDate(yearsToAdd) {
    const newDate = new Date();
    newDate.setHours(0, 0, 0, 0);
    newDate.setFullYear(newDate.getFullYear() + yearsToAdd);
    return newDate;
  },

  dateToSeconds(value, locale) {
    const parsedValue = value.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
    let getDate = parser.fromString(parsedValue, locale);
    if (getDate.invalid) {
      getDate = Date.parse(`${parsedValue} GMT`);
    }
    const formattedDate = getDate / 1000;
    return formattedDate;
  },
};
