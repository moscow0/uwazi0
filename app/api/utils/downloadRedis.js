/* eslint-disable no-console */
//eslint-disable-next-line node/no-restricted-import
import fs from 'fs';
import { execSync } from 'child_process';
import path from 'path';

export const downloadRedis = () => {
  const redisVersion = execSync('cat .redis_version').toString().replace('\n', '');
  const pathToBin = path.join(__dirname, '../../../redis-bin/redis-stable/src/redis-server');
  console.log(pathToBin);
  if (fs.existsSync(pathToBin)) {
    return;
  }
  console.log('Downloading redis...');

  execSync('mkdir -p redis-bin', { stdio: 'ignore' });
  execSync(
    `cd redis-bin && curl -O http://download.redis.io/releases/redis-${redisVersion}.tar.gz`,
    { stdio: 'ignore' }
  );
  execSync(`cd redis-bin && tar xzvf redis-${redisVersion}.tar.gz`, { stdio: 'ignore' });
  execSync(`cd redis-bin && mv redis-${redisVersion} redis-stable`, { stdio: 'ignore' });

  console.log('Downloading redis... Done');
  console.log('Installing redis...');
  execSync(
    `cd redis-bin &&
       cd redis-stable &&
       make distclean &&
       make`,
    { stdio: 'ignore' }
  );
  console.log('Installing redis... Done');
};
