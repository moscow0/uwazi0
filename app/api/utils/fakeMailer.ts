import Mail from 'nodemailer/lib/mailer';
import { SentMessageInfo } from 'nodemailer';

export class FakeMailer {
  // eslint-disable-next-line class-methods-use-this
  sendMail(
    _mailOptions: Mail.Options,
    callback: (err: Error | null, info: SentMessageInfo) => void
  ) {
    console.log('Fake sent of mail with:', _mailOptions);
    callback(null, '');
  }
}

export default {
  createTransport: (_transporter: string, _defaults: {}) => new FakeMailer(),
};
