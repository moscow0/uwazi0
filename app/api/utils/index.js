import validateRequest from './validateRequest';
import createError from './Error';

export { handleError } from './handleError';
export { parseQuery } from './parseQueryMiddleware';

const validation = {
  validateRequest,
};

export { validation, createError };
