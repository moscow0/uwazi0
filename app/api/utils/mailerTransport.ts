import nodemailer from 'nodemailer';
import fakeMailer from './fakeMailer';

export const getMailerTransport = () =>
  process.env.DATABASE_NAME !== 'uwazi_e2e' ? nodemailer : fakeMailer;
