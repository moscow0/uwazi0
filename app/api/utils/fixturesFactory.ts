import { ObjectId } from 'mongodb';
import db from 'api/utils/testing_db';
import { EntitySchema } from 'shared/types/entityType';
import { FileType } from 'shared/types/fileType';
import { UserRole } from 'shared/types/userSchema';
import { UserSchema } from 'shared/types/userType';
import { ThesaurusValueSchema } from 'shared/types/thesaurusType';
import {
  PropertySchema,
  MetadataSchema,
  PropertyValueSchema,
  MetadataObjectSchema,
  ExtractedMetadataSchema,
} from 'shared/types/commonTypes';
import { UpdateLog } from 'api/updatelogs';

function getIdMapper() {
  const map = new Map<string, ObjectId>();

  return function setAndGet(key: string) {
    if (!map.has(key)) map.set(key, db.id() as ObjectId);

    return map.get(key)!;
  };
}

const thesaurusNestedValues = (
  rootValue: string,
  children: Array<string>,
  idMapper: (key: string) => ObjectId
) => {
  const nestedValues = children.map(nestedValue => ({
    _id: idMapper(nestedValue),
    id: nestedValue,
    label: nestedValue,
  }));
  return { _id: idMapper(rootValue), id: rootValue, label: rootValue, values: nestedValues };
};

function getFixturesFactory() {
  const idMapper = getIdMapper();

  return Object.freeze({
    id: idMapper,

    template: (name: string, properties: PropertySchema[] = []) => ({
      _id: idMapper(name),
      name,
      properties,
    }),

    entity: (
      id: string,
      template?: string,
      metadata: MetadataSchema = {},
      props: EntitySchema = { language: 'en' }
    ): EntitySchema => {
      const language = props.language || 'en';
      return {
        _id: idMapper(`${id}-${language}`),
        sharedId: id,
        title: `${id}`,
        ...(template ? { template: idMapper(template) } : {}),
        metadata,
        language,
        ...props,
      };
    },

    inherit(name: string, content: string, property: string, props = {}): PropertySchema {
      return this.relationshipProp(name, content, {
        inherit: { property: idMapper(property).toString() },
        ...props,
      });
    },

    file: (
      id: string,
      entity: string | undefined,
      type: 'custom' | 'document' | 'thumbnail' | 'attachment' | undefined,
      filename: string,
      language: string = 'en',
      originalname: string | undefined = undefined,
      extractedMetadata: ExtractedMetadataSchema[] = []
    ): FileType => ({
      _id: idMapper(`${id}`),
      entity,
      language,
      type,
      filename,
      originalname: originalname || filename,
      extractedMetadata,
    }),

    relationshipProp(name: string, content: string, props = {}): PropertySchema {
      return this.property(name, 'relationship', {
        relationType: idMapper('rel1').toString(),
        content: idMapper(content).toString(),
        ...props,
      });
    },

    property: (
      name: string,
      type: PropertySchema['type'] = 'text',
      props = {}
    ): PropertySchema => ({
      _id: idMapper(name),
      label: name,
      name,
      type,
      ...props,
    }),

    metadataValue: (value: PropertyValueSchema): MetadataObjectSchema => ({
      value,
    }),

    thesauri: (name: string, values: Array<string | [string, string]>) => ({
      name,
      _id: idMapper(name),
      values: values.map(value =>
        typeof value === 'string'
          ? { _id: idMapper(value), id: value, label: value }
          : { _id: idMapper(value[0]), id: value[0], label: value[1] }
      ),
    }),

    nestedThesauri: (name: string, values: Array<{ [k: string]: Array<string> } | string>) => {
      const thesaurusValues = values.reduce(
        (accumulator: ThesaurusValueSchema[], item: { [k: string]: Array<string> } | string) => {
          const nestedItems =
            typeof item === 'string'
              ? [{ _id: idMapper(item), id: item, label: item }]
              : Object.entries(item).map(([rootValue, children]) =>
                  thesaurusNestedValues(rootValue, children, idMapper)
                );
          return [...accumulator, ...nestedItems];
        },
        []
      );
      return {
        name,
        _id: idMapper(name),
        values: thesaurusValues,
      };
    },

    user: (username: string, role?: UserRole, email?: string, password?: string): UserSchema => ({
      username,
      _id: idMapper(username),
      role: role || UserRole.COLLABORATOR,
      email: email || `${username}@provider.tld`,
      password,
    }),

    updatelog: (
      namespace: string,
      mongoId: string,
      deleted = false,
      timestamp = Date.now()
    ): Partial<UpdateLog> => ({
      _id: idMapper(`${namespace}-${mongoId}`),
      namespace,
      mongoId: idMapper(mongoId),
      timestamp,
      deleted,
    }),
  });
}

export { getIdMapper, getFixturesFactory };
