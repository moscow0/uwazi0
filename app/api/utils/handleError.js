import { errorLog, debugLog } from 'api/log';
import Ajv from 'ajv';
import { createError } from 'api/utils/index';
import { appContext } from 'api/utils/AppContext';

const ajvPrettifier = error => {
  const errorMessage = [error.message];
  if (error.errors && error.errors.length) {
    error.errors.forEach(oneError => {
      errorMessage.push(`${oneError.instancePath}: ${oneError.message}`);
    });
  }
  return errorMessage.join('\n');
};

const joiPrettifier = (error, req) => {
  const url = req.originalUrl ? `\nurl: ${req.originalUrl}` : '';
  const body =
    req.body && Object.keys(req.body).length
      ? `\nbody: ${JSON.stringify(req.body, null, ' ')}`
      : '';
  const query =
    req.query && Object.keys(req.query).length
      ? `\nquery: ${JSON.stringify(req.query, null, ' ')}`
      : '';
  const errorString = `\n${error.message || JSON.stringify(error.json)}`;

  let errorMessage = `${url}${body}${query}${errorString}`;

  //if the resulting message is empty, or meaningless combination of characters ('{}')
  if (errorMessage.match(/^[{}\s]*$/g)) {
    errorMessage = JSON.stringify(error, null, 2);
  }

  return errorMessage;
};

const appendOriginalError = (message, originalError) =>
  `${message}\noriginal error: ${JSON.stringify(originalError, null, ' ')}`;

const obfuscateCredentials = req => {
  const obfuscated = req;
  if (req.body && req.body.password) {
    obfuscated.body.password = '########';
  }

  if (req.body && req.body.username) {
    obfuscated.body.username = '########';
  }

  return obfuscated;
};

// eslint-disable-next-line max-statements
const prettifyError = (error, { req = {}, uncaught = false } = {}) => {
  let result = error;

  if (error instanceof Error) {
    result = { code: 500, message: error.stack };
  }

  if (error instanceof Ajv.ValidationError) {
    result = { code: 422, message: error.message, validations: error.errors };
  }

  if (error.name === 'ValidationError') {
    result = { code: 422, message: error.message, validations: error.properties };
  }

  if (error.name === 'MongoError') {
    result.code = 500;
  }

  if (error.message && error.message.match(/Cast to ObjectId failed for value/)) {
    result.code = 400;
  }

  if (error.message && error.message.match(/rison decoder error/)) {
    result.code = 400;
  }

  if (uncaught) {
    result.message = `uncaught exception or unhandled rejection, Node process finished !!\n ${result.message}`;
  }

  const obfuscatedRequest = obfuscateCredentials(req);
  result.prettyMessage = error.ajv
    ? ajvPrettifier(result)
    : joiPrettifier(result, obfuscatedRequest);

  return result;
};

const getErrorMessage = (data, error) => {
  const originalError = data.original || error;
  const prettyMessage = data.requestId
    ? `requestId: ${data.requestId} ${data.prettyMessage}`
    : data.prettyMessage;

  if (originalError instanceof Error) {
    const extendedError = appendOriginalError(prettyMessage, originalError);
    return data.tenantError
      ? `${extendedError}\n[Tenant error] ${data.tenantError.message}`
      : extendedError;
  }

  return prettyMessage;
};

const sendLog = (data, error, errorOptions) => {
  const messageToLog = getErrorMessage(data, error);
  if (data.code === 500) {
    errorLog.error(messageToLog, errorOptions);
  } else if (data.code === 400) {
    debugLog.debug(messageToLog, errorOptions);
  }
};

function simplifyError(result, error) {
  const simplifiedError = { ...result };
  delete simplifiedError.original;

  if (error instanceof Error) {
    simplifiedError.prettyMessage = error.message;
    simplifiedError.error = error.message;
    delete simplifiedError.message;
  } else {
    simplifiedError.prettyMessage = simplifiedError.prettyMessage || error.message;
  }

  return simplifiedError;
}

function setRequestId(result) {
  try {
    return { ...result, requestId: appContext.get('requestId') };
  } catch (err) {
    return { ...result, tenantError: err };
  }
}

const handleError = (_error, { req = undefined, uncaught = false, useContext = true } = {}) => {
  const errorData = typeof _error === 'string' ? createError(_error, 500) : _error;

  const error = errorData || new Error('Unexpected error has occurred');
  let result = prettifyError(error, { req, uncaught });

  if (useContext) {
    result = setRequestId(result);
  }

  sendLog(result, error, {});

  return simplifyError(result, error);
};

export { handleError, prettifyError };
