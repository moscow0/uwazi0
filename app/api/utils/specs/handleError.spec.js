import { createError } from 'api/utils';
import { errorLog, debugLog } from 'api/log';

import { errors as elasticErrors } from '@elastic/elasticsearch';
import { appContext } from 'api/utils/AppContext';
import { handleError, prettifyError } from '../handleError';

const contextRequestId = '1234';

const { ConnectionError } = elasticErrors;

describe('handleError', () => {
  beforeEach(() => {
    spyOn(errorLog, 'error');
    spyOn(debugLog, 'debug');
    spyOn(appContext, 'get').and.returnValue(contextRequestId);
  });

  describe('errors by type', () => {
    describe('when error is instance of Error', () => {
      it('should return the error with 500 code without the original error and error stack', () => {
        const errorInstance = new Error('error');
        const error = handleError(errorInstance);

        expect(error.code).toBe(500);
        expect(error.requestId).toBe(contextRequestId);
        expect(error.prettyMessage).toEqual('error');
        expect(error.message).toBeUndefined();
        expect(error.original).toBeUndefined();
      });

      it('should correctly log the original error for ElasticSearch exceptions', () => {
        const error = new ConnectionError('test error', { meta: 'some meta' });
        handleError(error);

        expect(errorLog.error).toHaveBeenCalledWith(
          `requestId: ${contextRequestId} \n${error.stack}
original error: {
 "name": "ConnectionError",
 "meta": {
  "meta": "some meta"
 }
}`,
          {}
        );
      });

      it('should log the error with the requestId', () => {
        const error = new Error('error');
        handleError(error);

        expect(errorLog.error).toHaveBeenCalledWith(
          `requestId: ${contextRequestId} \n${error.stack}\noriginal error: {}`,
          {}
        );
      });
    });

    describe('when error is created with createError', () => {
      it('should return the error', () => {
        const error = handleError(createError('test error', 400));
        expect(error).toMatchSnapshot();
      });

      it('should not log the error when code is not 500', () => {
        handleError(createError('test error', 400));
        expect(errorLog.error).not.toHaveBeenCalled();

        handleError(createError('test error'));
        expect(errorLog.error).toHaveBeenCalledWith(
          `requestId: ${contextRequestId} \ntest error`,
          {}
        );
      });
    });

    describe('when error is a MongoError', () => {
      it('should return the error with a 500 code', () => {
        const error = handleError({ name: 'MongoError', message: 'error', code: '345' });
        expect(error.code).toBe(500);
        expect(error.message).toBe('error');
      });
    });

    describe('when error is a mongoose ValidationError', () => {
      it('should return the error with a 422 error', () => {
        const error = handleError({ name: 'ValidationError', message: 'error', code: '1000' });
        expect(error.code).toBe(422);
        expect(error.message).toBe('error');
      });
    });
  });

  describe('when error is undefined', () => {
    it('should return generate a new error with code 500', () => {
      const error = handleError();
      expect(error.code).toBe(500);
      expect(error.prettyMessage).toMatch(/Unexpected error has occurred/i);
      expect(error.requestId).toBe(contextRequestId);
    });
  });

  describe('when error is uncaught', () => {
    it('should append the info into the message', () => {
      const uncaught = true;
      const error = handleError({ message: 'error' }, { uncaught });
      expect(error.message).toBe(
        'uncaught exception or unhandled rejection, Node process finished !!\n error'
      );
    });
  });

  describe('when "Cast to objectId failed"', () => {
    it('should set code to 400', () => {
      const error = handleError({ message: 'Cast to ObjectId failed for value' });
      expect(error.code).toBe(400);
    });
  });

  describe('when "rison decoder error"', () => {
    it('should set code to 400', () => {
      const error = handleError({ message: 'rison decoder error' });
      expect(error.code).toBe(400);
    });
  });

  describe('when error is 400', () => {
    it('should log it using debugLog', () => {
      handleError(createError('test error', 400));
      expect(debugLog.debug.calls.mostRecent().args[0]).toContain('test error');
    });

    describe('and is instance of Error', () => {
      it('should include the original error', () => {
        const error = new Error('test error');
        error.name = 'Original error';
        handleError(createError(error, 400));
        expect(debugLog.debug.calls.mostRecent().args[0]).toContain('Original error');
      });
    });
  });

  describe('when the body contains the user and password', () => {
    it('should not show them in the log', () => {
      handleError(createError('test error', 400), {
        req: { body: { username: 'admin', password: '1234' } },
      });
      expect(debugLog.debug.calls.allArgs()).toMatchSnapshot();
    });
  });

  describe('when error has ajv validation errors', () => {
    it('should prettify the message', () => {
      const errorInstance = {
        ajv: true,
        code: 400,
        message: 'hello',
        errors: [
          {
            instancePath: 'a property',
            message: 'an error',
          },
        ],
      };
      const error = handleError(errorInstance);

      expect(error.code).toBe(400);
      const expectedPrettymessage = 'hello\na property: an error';
      expect(error.prettyMessage).toEqual(expectedPrettymessage);
    });
  });
});

describe('handleError without context', () => {
  it('should append a tenant error message to the original error', () => {
    spyOn(errorLog, 'error');
    const error = handleError(new Error('original error message'));
    expect(error.prettyMessage).toEqual('original error message');
    expect(errorLog.error).toHaveBeenCalledWith(
      expect.stringMatching(
        /\nError: original error message[\w\W]*Accessing nonexistent async context/
      ),
      {}
    );
  });
});

describe('prettifyError', () => {
  describe('when the error does not fall into any other category, and the resulting message would be empty', () => {
    it('should return JSON representation of the original error object as a message.', () => {
      const prettied = prettifyError({
        json: {},
        status: '404',
        headers: 'some_headers',
      });

      expect(prettied.prettyMessage).toBe(
        '{\n  "json": {},\n  "status": "404",\n  "headers": "some_headers"\n}'
      );
    });
  });
});
