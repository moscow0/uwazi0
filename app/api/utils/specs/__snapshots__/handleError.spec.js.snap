// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleError errors by type when error is created with createError should return the error 1`] = `
Object {
  "code": 400,
  "message": "test error",
  "prettyMessage": "
test error",
  "requestId": "1234",
}
`;

exports[`handleError when the body contains the user and password should not show them in the log 1`] = `
Array [
  Array [
    "requestId: 1234 
body: {
 \\"username\\": \\"########\\",
 \\"password\\": \\"########\\"
}
test error",
    Object {},
  ],
]
`;
